networking/nat/src/main/java/tech/pegasys/teku/networking/nat/NatManager.java

```diff
@@ -0,0 +1,423 @@
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.networking.nat;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static tech.pegasys.teku.infrastructure.async.FutureUtil.ignoreFuture;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jupnp.UpnpService;
+import org.jupnp.UpnpServiceImpl;
+import org.jupnp.model.action.ActionInvocation;
+import org.jupnp.model.message.UpnpResponse;
+import org.jupnp.model.meta.RemoteDevice;
+import org.jupnp.model.meta.RemoteDeviceIdentity;
+import org.jupnp.model.meta.RemoteService;
+import org.jupnp.model.types.UnsignedIntegerFourBytes;
+import org.jupnp.model.types.UnsignedIntegerTwoBytes;
+import org.jupnp.registry.Registry;
+import org.jupnp.registry.RegistryListener;
+import org.jupnp.support.igd.callback.GetExternalIP;
+import org.jupnp.support.igd.callback.PortMappingAdd;
+import org.jupnp.support.igd.callback.PortMappingDelete;
+import org.jupnp.support.model.PortMapping;
+import tech.pegasys.teku.infrastructure.async.FutureUtil;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.service.serviceutils.Service;
+
+@SuppressWarnings("unchecked")
+public class NatManager extends Service {
+  protected static final Logger LOG = LogManager.getLogger();
+
+  static final String SERVICE_TYPE_WAN_IP_CONNECTION = "WANIPConnection";
+
+  private final UpnpService upnpService;
+  private final RegistryListener registryListener;
+  private final SafeFuture<String> externalIpQueryFuture = new SafeFuture<>();
+
+  private final Map<String, SafeFuture<RemoteService>> recognizedServices = new HashMap<>();
+  private final List<NatPortMapping> forwardedPorts = new ArrayList<>();
+  private Optional<String> discoveredOnLocalAddress = Optional.empty();
+
+  public NatManager(final NatMethod natMethod) {
+    // this(new UpnpServiceImpl(new DefaultUpnpServiceConfiguration()));
+
+    // Workaround for an issue in the jupnp library: the ExecutorService used misconfigures
+    // its ThreadPoolExecutor, causing it to only launch a single thread. This prevents any work
+    // from getting done (effectively a deadlock). The issue is fixed here:
+    //   https://github.com/jupnp/jupnp/pull/117
+    // However, this fix has not made it into any releases yet.
+    // TODO: once a new release is available, remove this @Override
+    this(new UpnpServiceImpl(new TekuNatServiceConfiguration()), natMethod);
```

Comment:
Can we work out how to disable the `SpecificationViolationReporter` - it winds up being very noisy:

```
2021-01-27 11:04:35.366+10:00 | jupnp-16 | WARN  | SpecificationViolationReporter | UPnP specification violation: Unsupported UDA minor version, ignoring: 1
2021-01-27 11:04:35.370+10:00 | jupnp-8 | WARN  | SpecificationViolationReporter | UPnP specification violation: UPC must be 12 digits: 'Linux IGD' for device 'WANConnectionDevice'
2021-01-27 11:04:35.370+10:00 | jupnp-8 | WARN  | SpecificationViolationReporter | UPnP specification violation: UPC must be 12 digits: 'Linux IGD' for device 'WANDevice'
2021-01-27 11:04:35.370+10:00 | jupnp-8 | WARN  | SpecificationViolationReporter | UPnP specification violation: UPC must be 12 digits: 'Linux IGD' for device 'WANConnectionDevice'
2021-01-27 11:04:35.371+10:00 | jupnp-8 | WARN  | SpecificationViolationReporter | UPnP specification violation: UPC must be 12 digits: 'Linux IGD' for device 'WANDevice'
2021-01-27 11:04:35.371+10:00 | jupnp-8 | WARN  | SpecificationViolationReporter | UPnP specification violation: UPC must be 12 digits: 'Linux IGD' for device 'WANConnectionDevice'
2021-01-27 11:04:35.377+10:00 | jupnp-10 | WARN  | SpecificationViolationReporter | UPnP specification violation: Unsupported UDA minor version, ignoring: 1
2021-01-27 11:04:35.411+10:00 | jupnp-8 | WARN  | SpecificationViolationReporter | UPnP specification violation: UPC must be 12 digits: 'Linux IGD' for device 'WANConnectionDevice'
2021-01-27 11:04:35.418+10:00 | jupnp-8 | WARN  | SpecificationViolationReporter | UPnP specification violation: UPC must be 12 digits: 'Linux IGD' for device 'WANDevice'
2021-01-27 11:04:35.424+10:00 | jupnp-8 | WARN  | SpecificationViolationReporter | UPnP specification violation: UPC must be 12 digits: 'Linux IGD' for device 'WANConnectionDevice'
2021-01-27 11:04:35.439+10:00 | jupnp-8 | WARN  | SpecificationViolationReporter | UPnP specification violation: UPC must be 12 digits: 'Linux IGD' for device 'WANDevice'
2021-01-27 11:04:35.446+10:00 | jupnp-8 | WARN  | SpecificationViolationReporter | UPnP specification violation: UPC must be 12 digits: 'Linux IGD' for device 'WANConnectionDevice'
```

Reply:
yeh that was annoying me too. i think we should be able to do it through logger configuration, i'll have a look.

Reply:
Looks like `SpecificationViolationReporter.disableReporting()` is probably all that's required.  It's a singleton - icky but convenient in this case...